keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,535 @@
+will be able to load env vars from a file.
+
+### Non-Goals
+
+1. We do not plan to support expansion of environment variables, that is, when an
+environment variable refers to another.
+
+2. We do not intend to update env vars if the file content has changed after the
+container consuming it has started.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+The user is using a container offered by a vendors that requires configuring
+env vars. [Some
+vendors](https://developer.hashicorp.com/vault/docs/platform/k8s/injector/examples#environment-variable-example)
+have to couple their app with Kubernetes to allow setting env vars in their
+container.
+
+Instead, k8s can provide a new mechanism that will be used to instantiate
+env vars from the specified file. The user can generate an env file using an
+`initContainer`. The container can then use the `fileKeyRef` field to reference
+the file from which it should initialize the env vars. Behind the scene, kubelet will
+parse the env file and instantiate these env vars during container creation. It
+does not mount this file onto the main container. In the example below,
+the Pod's output includes `CONFIG_VAR=HELLO`.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: dapi-test-pod
+spec:
+  initContainers:
+  - name: setup-envfile
+    image: registry.k8s.io/busybox
+    command: ['sh', '-c', 'echo "CONFIG_VAR=HELLO" > /data/config.env']
+    volumeMounts:
+    - name: config
+      mountPath: /data
+   containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    env:
+    - name: CONFIG_VAR
+      valueFrom:
+        fileKeyRef:
+          path: config.env
+          volumeName: config
+          key: CONFIG_VAR
+  restartPolicy: Never
+  volumes:
+  - name: config
+    emptyDir: {}
+```
+
+#### Story 2
+
+The user's workload needs to consume secrets of a service. Usually, they can
+store it in k8s Secret resource and use it to instantiate the env var. But it
+can become difficult to manage if each Pod needs to consume unique secrets.
+
+Instead, k8s can provide a new mechanism to fetch secrets in initContainer and
+store it in a file. The container can then use the `fileKeyRef` field to
+instantiate the env var from the file. Behind the scene, kubelet will parse
+the env file and instantiate these env vars during container creation. It
+does not mount this file onto the main container. In the example below,
+the Pod's output includes `JWT=token`.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: dapi-test-pod
+spec:
+  initContainers:
+  - name: setup-envfile
+    image: registry.k8s.io/busybox
+    command: ['sh', '-c', 'echo "JWT=token" > /data/config.env']
+    volumeMounts:
+    - name: config
+      mountPath: /data
+   containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    env:
+    - name: CONFIG_VAR
```

Comment:
This example is even more confusing.  I'd expect the container's output to include `CONFIG_VAR=token` but the above test says it will retain the JWT name - so what is this name for?